stout/take.h

```diff
@@ -0,0 +1,194 @@
+      stream_ = &stream;
+      k_.Start(*this);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    template <typename... Args>
+    void Body(Args&&... args) {
+      if (data_.size() == n_) {
+        data_.pop_front();
+      }
+      data_.push_back(std::forward<Args>(args)...);
+      stream_->Next();
+    }
+
+    // should be called from the previous eventual (Stream)
+    // when the Stream is done, so we are ready to start
+    // streaming last values
+    void Ended() {
+      n_ = std::min(n_, data_.size());
+      endedTriggered_ = true;
+
+      auto value = data_.front();
+      data_.pop_front();
+      k_.Body(std::move(value));
+    }
+
+    void Next() override {
+      // When Next called from the next eventual,
+      // the element should be taken from the stored stream
+
+      // There are some values in the previous stream before we call
+      // k.Ended() on the stream
+      if (!endedTriggered_) {
+        stream_->Next();
+      } else {
+        if (data_.empty()) {
+          // When there are no more stored values in our new stream data
+          // stream have to be ended there
+          k_.Ended();
+        } else {
+          auto value = data_.front();
+          data_.pop_front();
+          k_.Body(std::move(value));
+        }
+      }
+    }
+
+    void Done() override {
+      k_.Ended();
+    }
+
+    K_ k_;
+    size_t n_;
+    std::deque<Arg_> data_;
+    bool endedTriggered_ = false;
+
+    TypeErasedStream* stream_ = nullptr;
+  };
+
+  struct Composable {
+    template <typename Arg>
+    using ValueFrom = Arg;
+
+    template <typename Arg, typename K>
+    auto k(K k) && {
+      return Continuation<K, Arg>{std::move(k), std::move(n_)};
+    }
+
+    size_t n_;
+  };
+};
+
+////////////////////////////////////////////////////////////////////////
+
+struct _TakeRange {
+  template <typename K_, typename Arg_>
+  struct Continuation {
+    void Start(TypeErasedStream& stream) {
+      stream_ = &stream;
+      k_.Start(stream);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    template <typename... Args>
+    // inRange_ needs to prevent calling Next
+    // when stream has already passed the set amount_ of elements
+    void Body(Args&&... args) {
+      if (CheckRange()) {
+        inRange_ = true;
+        k_.Body(std::forward<Args>(args)...);
+      } else if (!inRange_) {
+        stream_->Next();
+      } else {
+        stream_->Done();
+      }
+    }
+
+    void Ended() {
+      k_.Ended();
+    }
+
+    bool CheckRange() {
+      bool result = i_ >= begin_ && i_ < begin_ + amount_;
+      ++i_;
+      return result;
+    }
+
+    K_ k_;
+    size_t begin_;
+    size_t amount_;
+    size_t i_ = 0;
+    bool inRange_ = false;
```

Comment:
I believe our naming convention for variables is to use underscores between words, so:
```suggestion
    bool in_range_ = false;
```

[Here's an example](https://github.com/3rdparty/stout-grpc/blob/fef528dea479b09dec7bc49aa7535c0e87bca310/stout/grpc/server.h#L105) of where we already do that. Maybe Ben is following the [Google Style Guide](https://google.github.io/styleguide/cppguide.html#Variable_Names). 😃